libsigner/src/signer_set.rs

```diff
@@ -0,0 +1,137 @@
+        let mut wsts_key_ids = HashMap::new();
+        for (i, entry) in reward_set.iter().enumerate() {
+            let signer_id = u32::try_from(i).map_err(|_| Error::SignerCountOverflow)?;
+            let ecdsa_pk =
+                ecdsa::PublicKey::try_from(entry.signing_key.as_slice()).map_err(|e| {
+                    Error::BadSignerPublicKey(format!(
+                        "Failed to convert signing key to ecdsa::PublicKey: {e}"
+                    ))
+                })?;
+            let signer_public_key = Point::try_from(&Compressed::from(ecdsa_pk.to_bytes()))
+                .map_err(|e| {
+                    Error::BadSignerPublicKey(format!(
+                        "Failed to convert signing key to wsts::Point: {e}"
+                    ))
+                })?;
+            let stacks_public_key = StacksPublicKey::from_slice(entry.signing_key.as_slice())
+                .map_err(|e| {
+                    Error::BadSignerPublicKey(format!(
+                        "Failed to convert signing key to StacksPublicKey: {e}"
+                    ))
+                })?;
+
+            let stacks_address = StacksAddress::p2pkh(is_mainnet, &stacks_public_key);
+            signer_ids.insert(stacks_address, signer_id);
+
+            signer_public_keys.insert(signer_id, signer_public_key);
+            let weight_start = weight_end;
+            weight_end = weight_start + entry.weight;
+            let key_ids: HashSet<u32> = (weight_start..weight_end).collect();
+            for key_id in key_ids.iter() {
+                wsts_key_ids.insert(*key_id, ecdsa_pk.clone());
+            }
+            signer_key_ids.insert(signer_id, (weight_start..weight_end).collect());
+            coordinator_key_ids.insert(signer_id, key_ids);
+            wsts_signers.insert(signer_id, ecdsa_pk);
+        }
+
+        Ok(Self {
+            signer_ids,
+            public_keys: PublicKeys {
+                signers: wsts_signers,
+                key_ids: wsts_key_ids,
+            },
+            signer_key_ids,
+            signer_public_keys,
+            coordinator_key_ids,
+        })
+    }
+
+    /// Return the number of Key IDs in the WSTS group signature
+    pub fn count_keys(&self) -> Result<u32, Error> {
+        self.public_keys
+            .key_ids
+            .len()
+            .try_into()
+            .map_err(|_| Error::SignerCountOverflow)
+    }
+
+    /// Return the number of Key IDs in the WSTS group signature
+    pub fn count_signers(&self) -> Result<u32, Error> {
+        self.public_keys
+            .signers
+            .len()
+            .try_into()
+            .map_err(|_| Error::SignerCountOverflow)
+    }
+
+    /// Return the number of Key IDs required to sign a message with the WSTS group signature
+    pub fn get_signing_threshold(&self) -> Result<u32, Error> {
+        let num_keys = self.count_keys()?;
+        Ok((num_keys as f64 * 7_f64 / 10_f64).ceil() as u32)
```

Comment:
Should we use consistent math here?  As in, `num_keys * 700 / 1000`?

Reply:
Are floats not consistent here? Either way, these are the existing calculations from the signer side.

Reply:
If we want to be really sure of consistency, there is a readonly function, `get-threshold-weight`,  in the signers-voting clarity contract that you can call.

Reply:
The reason we're using floats here is that we need to round these calculations up, and integer math will silently round down.  We need to round up because if say 70% of the number of keys contains a fraction, then rounding down means that our threshold is less than 70%.

I know floats are not particularly consistent in general but for this calculation they should return a consistent answer across all platforms and environments.